rfd/0059-search-based-access-requests.md

```diff
@@ -0,0 +1,276 @@
+    # as a part of search_as request and in addition to that will be used to
+    # evaluate access.
+    db_labels:
+       owner: "{{params.access_group}}"
+```
+
+### Certificate issuance and RBAC
+
+Once a user assumes “db-admin-role” the Teleport root cluster issues a cert
+including the assumed role and a list of UUIDs of resources granted by the
+access request.
+
+This makes sure that the role is scoped to a static list of resources that never
+changes for this certificate.
+
+```
+Assumed-role: [db-admin-role]
+Resource-UUIDs: [uuid-1, uuid-2]
+```
+
+### Trusted clusters
+
+Leaf clusters will use standard role mapping to validate the cert issued by the
+root. If the leaf cluster role maps root cluster’s “db-admin-role” to the same role
+using cluster mapping:
+
+```yaml
+role_map:
+   '*': '*'
+```
+
+Then leaf cluster behavior will be identical. Leaf cluster may choose to narrow
+the scope, for example:
+
+```yaml
+role_map:
+   'db-admin-role': 'leaf-db-admin-role'
+```
+
+```yaml
+kind: role
+metadata:
+  name: leaf-db-admin-role
+spec:
+  # This role accepts a parameter "access_group". This parameter
+  # is either supplied when evaluating the role during access check
+  # and it should be present in the cert, or it is passed as a local
+  # variable as a part of "search-as"
+  params:
+   - name: access_group
+  allow:
+    logins: ["root"]
+    # node_labels defines what nodes this role will be allowed to search
+    # as a part of search_as request and in addition to that will be used to
+evaluate access
+    node_labels:
+       owner: "{{params.access_group}}"
+       class: external-access-allowed
+```
+
+In this case, if the cert issued by the root cluster granted access to nodes
+with uuid “uuid-1” and “uuid-2”. In case if leaf’s uuid-1 has label “class:
+external-access-allowed” and uuid-2 does not. Leaf cluster will reject access to
+the node uuid-2 despite the fact that the root cluster “allowed” it.
+
+### Access Requests RBAC Edge-Cases
+
+A user could find and request access to several different types of resources,
+some with `label: a` others with `label: b`.
+Access to resource `label: a` can be granted by users having role `a`.
+Access to resource `label: b` can be granted by users having role: `b`.
+In this case both users with role `a` and role `b` have to approve access for
+the request to be granted.
+
+### Access Requests RBAC Plugin Notification
+
+Access request slack and mattermost plugin should add routing to channels by
+label and resource type.
+
+Will complete this section with an example config, or create another RFD.
+
+### Audit events
+
+Will create a new audit event `access_request.search` that will include the searches users are
+running in the scope of requesting access.
+
+### OpenSSH
+
+OpenSSH clients should work with credentials obtained via search-based access requests, assuming that the proxy and node are both Teleport.
+OpenSSH servers won't support this flow.
```

Comment:
We can add `-P` flag to proxy command's `tsh proxy ssh -P` as a proxy command for OpenSSH to work